man/tpm2_clearlock.1.md
% tpm2_clearlock(1) tpm2-tools | General Commands Manual
%
% DECEMBER 2017

# NAME

**tpm2_clearlock**(1) - Lock/unlock access to the clear operation.

# SYNOPSIS

**tpm2_clearlock** [OPTIONS]

# DESCRIPTION

**tpm2_clearlock**(1) - Allow a user to enable (unlock) or disable (lock)
access to the **tpm2_clear** operation. If the lockout password option
is missing, assume NULL.

# OPTIONS

  * **-c**, **\--clear**:

    Specifies the tool should unlock access to the clear command.
    By default it will try to disable the clear command.

  * **-p**, **\--platform**:

    Specifies the tool should operate on the platform hierarchy. By default
    it operates on the lockout hierarchy.

    **NOTE : Operating on platform hierarchy require platform authentication.**

  * **-L**, **\--auth-lockout**=_LOCKOUT\_PASSWORD_:

    The lockout authorization value.

    Authorization values should follow the "authorization formatting standards",
    see section "Authorization Formatting".
    This tool only respects the *Password* and *HMAC* options.

[common options](common/options.md)

[common tcti options](common/tcti.md)

[authorization formatting](common/authorizations.md)

# EXAMPLES

## Enable the clear command on the platform hierarchy
```
tpm2_clearlock -c -p -L lockoutpasswd
```

## Disable the clear command on the lockout hierarchy
```
tpm2_clearlock
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_clearlock.c
/* SPDX-License-Identifier: BSD-3-Clause */
/*
 * Copyright (c) 2017, Emmanuel Deloget <logout@free.fr>
 * Copyright (c) 2018, Intel Corporation
 * All rights reserved.
 *
 */
#include <stdbool.h>
#include <stdlib.h>

#include "log.h"
#include "tpm2_auth_util.h"
#include "tpm2_tool.h"
#include "tpm2_util.h"

typedef struct clearlock_ctx clearlock_ctx;
struct clearlock_ctx {
    bool clear;
    bool platform;
    TPMS_AUTH_COMMAND session_data;
    struct {
        UINT8 L : 1;
        UINT8 unused : 7;
    } flags;
    char *lockout_auth_str;
};

static clearlock_ctx ctx = {
    .clear = false,
    .platform = false,
    .session_data = TPMS_AUTH_COMMAND_INIT(TPM2_RS_PW),
    .flags = { 0 },
};

static bool clearlock(ESYS_CONTEXT *ectx) {

    ESYS_TR rh = ctx.platform ? ESYS_TR_RH_PLATFORM : ESYS_TR_RH_LOCKOUT;
    TPMI_YES_NO disable = ctx.clear ? 0 : 1;

    LOG_INFO ("Sending TPM2_ClearControl(%s) command on %s",
            ctx.clear ? "CLEAR" : "SET",
            ctx.platform ? "TPM2_RH_PLATFORM" : "TPM2_RH_LOCKOUT");

    TSS2_RC rval = Esys_TR_SetAuth(ectx, rh, &ctx.session_data.hmac);
    if (rval != TPM2_RC_SUCCESS) {
        LOG_PERR(Esys_TR_SetAuth, rval);
        return false;
    }
    
    rval = Esys_ClearControl(ectx, rh,
                ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE, disable);
    if (rval != TPM2_RC_SUCCESS && rval != TPM2_RC_INITIALIZE) {
        LOG_PERR(Esys_ClearControl, rval);
        return false;
    }

    LOG_INFO ("Success. TSS2_RC: 0x%x", rval);
    return true;
}

static bool on_option(char key, char *value) {

    switch (key) {
    case 'c':
        ctx.clear = true;
        break;
    case 'p':
        ctx.platform = true;
        break;
    case 'L':
        ctx.flags.L = 1;
        ctx.lockout_auth_str = value;
        break;
    }

    return true;
}

bool tpm2_tool_onstart(tpm2_options **opts) {

    const struct option topts[] = {
        { "clear",          no_argument,       NULL, 'c' },
        { "auth-lockout",   required_argument, NULL, 'L' },
        { "platform",       no_argument,       NULL, 'p' },
    };

    *opts = tpm2_options_new("cL:p", ARRAY_LEN(topts), topts,
                             on_option, NULL, 0);

    return *opts != NULL;
}

int tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    if (ctx.flags.L) {
        bool res = tpm2_auth_util_from_optarg(ectx, ctx.lockout_auth_str,
                &ctx.session_data, NULL);
        if (!res) {
            LOG_ERR("Invalid lockout authorization, got\"%s\"",
                ctx.lockout_auth_str);
            return 1;
        }
    }

    return clearlock(ectx) != true;
}
